### PR TITLE
buffer barrier channel

### PR DESCRIPTION
**Pull request by Stebalien**
Otherwise, we'll leak a goroutine if the user has walked away.

Also, make sure canceling the context actually cancels the barrier (test).

### CODE REVIEW

File: sdk/sync/redis_test.go [hunk 1]
```diff
@@ -140,6 +140,32 @@ func TestBarrier(t *testing.T) {
...
+	select {
+	case err := <-ch:
+		if err != context.Canceled {
+			t.Errorf("expected context canceled, got %s", err)
+		}
+	case <-time.After(3 * time.Second):
+		t.Error("expected a cancel")
+		return
+	}
+}
```
Comment (Stebalien) [inline sdk/sync/redis_test.go]: This _doesn't_ test the change in this PR. I just added this test to make sure barrier cancellation actually works.

Comment (Stebalien) [reply sdk/sync/redis_test.go]: Ok, so this _is_ actually broken. It looks like redis treats context like a suggestion, more than anything else.

Comment (Stebalien) [reply sdk/sync/redis_test.go]: Nevermind, my test is broken.

File: sdk/sync/redis_test.go [hunk 2]
```diff
@@ -140,6 +140,35 @@ func TestBarrier(t *testing.T) {
...
+
+	state := State("yoda")
+	ch := watcher.Barrier(ctx, state, 10)
+	cancel()
+	select {
+	case err := <-ch:
+		if err == nil {
+			t.Errorf("expected an error")
+		}
+	case <-time.After(3 * time.Second):
```
Comment (nonsense) [inline sdk/sync/redis_test.go]: Not sure why this timeout is after 3 seconds, when the context is set to 5 seconds timeout.

Comment (nonsense) [reply sdk/sync/redis_test.go]: Oh now I get it, you actually manually call cancel, so this is not at all related with the timeout from above.

**Pull request merged**